### ISSUE TITLE
correct usage of AddSearchPath in IFileSystem

### ISSUE THREAD

**Issue by tschumann** [ValveSoftware/halflife#1775]
I'm trying to mount content from Opposing Force and this is what I've tried:
g_pFileSystem->AddSearchPath( "gearbox", "GAME" ); - can't find models/blkop_apache.mdl
g_pFileSystem->AddSearchPath( "gearbox\\gearbox", "GAME" ); - can't find models/blkop_apache.mdl
g_pFileSystem->AddSearchPath( "gearbox\\models", "GAME" ); - can't find models/blkop_apache.mdl or blkop_apache.mdl
g_pFileSystem->AddSearchPath( "gearbox\\gearbox\\models", "GAME" );  - can't find models/blkop_apache.mdl or blkop_apache.mdl
I had some combination of things where it could find blkop_apache.mdl but I can't remember how I did that now. But that isn't ideal anyway, I want to be able to use the correct path.
g_pFileSystem->GetCurrentDirectory returns steamapps/common/Half-Life (plus the stuff at the front too).

Has anyone successfully used AddSearchPath, and what is the right syntax to get this to work?

**Comment by SamVanheer** [comment]
Where exactly did you put the code that adds the search path?

I put it in GameDLLInit, specifically here: https://github.com/SamVanheer/HLEnhanced/blob/693ed1401d2df00278061d32e509609f58be1323/game/shared/CBaseGameInterface.cpp#L82

```
g_pFileSystem = ( IFileSystem* ) fileSystemFactory( FILESYSTEM_INTERFACE_VERSION, nullptr );
assert( g_pFileSystem );
if( !g_pFileSystem )
{
	return false;
}

g_pFileSystem->AddSearchPath( "gearbox", "GAME" );

return true;
```

Then i just replaced the Apache's precache and setmodel calls with that model.

You'll need to set it on the client regardless and on the server if it's a dedicated server since they share the filesystem.

**Comment by tschumann** [comment]
I'm calling g_pFileSystem->AddSearchPath in W_Precache but I'm not doing it client-side, so I guess I'll need to add a call to g_pFileSystem->AddSearchPath in client.dll too (at https://github.com/SamVanheer/HLEnhanced/blob/8cd53f51e2c5ea33c0f1a2d6b5c57ac98083fd49/game/client/cdll_int.cpp#L134 I guess).
Semi-related - is cl_enginefuncs_s.COM_AddAppDirectoryToSearchPath a pass-through to AddSearchPath or is it something different?

Many thanks anyway.

**Comment by SamVanheer** [comment]
W_Precache is called every map start so that's not a great idea. It should ideally be called once only on startup.

COM_AddAppDirectoryToSearchPath does this:
```
void COM_AddAppDirectoryToSearchPath(const char *pszBaseDir)
{
  COM_AddAppDirectory((char *)pszBaseDir);
}

void COM_AddAppDirectory(char *pszBaseDir)
{
  FS_AddSearchPath(pszBaseDir, "PLATFORM");
}

void FS_AddSearchPath(const char *pPath, const char *pathID)
{
  g_pFileSystem->AddSearchPath(g_pFileSystem, pPath, pathID);
}
```

The "PLATFORM" path ID won't be searched by hw.dll code explicitly, so unless it's searching through all paths it won't find it. It should do that when looking for models though.

I'd put the search path adding code in a helper function that both the client and server can call. Always call it on the client, call it on the server if g_engfuncs.pfnIsDedicatedServer() returns true.

**Comment by tschumann** [comment]
Ah good point, thanks - I'll give that a go.

**Comment by tschumann** [comment]
Okay, so I'm still not having any luck with this.

On the the client I have:
`
void CL_LoadFileSystem( void )
{
	char szPDir[512];

	if ( gEngfuncs.COM_ExpandFilename( FILESYSTEM_DLLNAME, szPDir, sizeof( szPDir ) ) == FALSE )
	{
		ConsoleDPrintf( "Unable to load %s\n", FILESYSTEM_DLLNAME );
		g_hFilesystemModule = NULL;
		return;
	}

	g_hFilesystemModule = Sys_LoadModule( szPDir );
	CreateInterfaceFn filesystemFactory = Sys_GetFactory( g_hFilesystemModule );

	if ( g_hFilesystemModule == NULL )
	{
		ConsoleDPrintf( "Unable to get factory from %s\n", g_hFilesystemModule );
		g_hFilesystemModule = NULL;
		return;
	}

	g_pFilesystem = (IFileSystem *)filesystemFactory( FILESYSTEM_INTERFACE_VERSION, NULL);

	if ( g_pFilesystem )
	{
		ConsoleDPrintf( "%s interface instantiated\n", FILESYSTEM_INTERFACE_VERSION );
		g_pFilesystem->AddSearchPath("gearbox", "GAME");
	}
}
`
I call it from Initialize (I get no console output but if I call it from HUD_Init I see VFileSystem009 interface instantiated get printed).

On the server I have:
`
void LoadFileSystem( void )
{
	hFileSystemModule = Sys_LoadModule( FILESYSTEM_DLLNAME );
	CreateInterfaceFn fileSystemFactory = Sys_GetFactory( hFileSystemModule );

	if( fileSystemFactory == NULL )
	{
		g_pFileSystem = NULL;
	}

	g_pFileSystem = (IFileSystem *)fileSystemFactory( FILESYSTEM_INTERFACE_VERSION, NULL );

	if( g_pFileSystem )
	{
		ALERT( at_aiconsole, "%s interface instantiated.\n", FILESYSTEM_INTERFACE_VERSION );
	}
}
`
I call it via GameDLLInit and I can see VFileSystem009 interface instantiated get printed, and I call g_pFileSystem->AddSearchPath("gearbox", "GAME"); after calling LoadFileSystem.
I've tried calling PRECACHE_MODEL via W_Precache and in a monster's Spawn method but it still doesn't load blkop_apache.md;
I've also got fallback_dir in my liblist.gam but removing it doesn't help.
My code is at https://github.com/tschumann/basis

**Comment by SamVanheer** [comment]
Are you using the Steam HL engine or Xash? Are you sure it's loading the filesystem_stdio library used by the engine?

When i add the search path it works just fine when precaching `models/blkop_apache.mdl`.

Do note that you normally shouldn't precache anything in Spawn itself, though Precache is called from it.

Also, your code will try to call a null function pointer if fileSystemFactory is null on the server side.

If the fallback_dir doesn't work either then it might indicate a bigger problem. Does `gearbox/models/blkop_apache.mdl` actually exist? Is the `gearbox` directory located in the same directory as your mod directory?

**Comment by tschumann** [comment]
I'm using Steam, but I was using FileSystem_Steam.dll (which I guess is the problem). The paths are right and the file exists too.
And yeah you're right, I was moving stuff around trying to debug and didn't realise g_pFileSystem could be NULL.
I think fallback_dir was working I just mentioned it as a possible reason for things not working.

Anyway, thanks for the help so far - much appreciated.

**Comment by SamVanheer** [comment]
FileSystem_Steam is obsolete since it was used to read from Steam's storage formats (GCF). New installations don't have it anymore. Only StdIo is used by the engine, so if you use that it should work.

**Comment by tschumann** [comment]
So it looks like it loads okay without the client-side changes and if I add the dedicated server check on the server-side then it can't find the models. But that's okay, it's working.
I was going to ask about checking whether the client owns the game but seeing as the client check apparently doesn't matter I guess it's not relevant (not sure if was that doable either - I'm not sure what GetPlayerUniqueID returns - I got a jibberish string but I assume it's a Steam ID - but it looks like it only returns something once the player is in-game which is too late).

**Comment by tschumann** [comment]
Oh, and I'm guessing that mod code doesn't run before GameUI loads, so there's no way to have the Create Server dialogue contain maps from mounted mods?

**Comment by SamVanheer** [comment]
GetPlayerUniqueID is the player's hashed CD key, not the Steam ID.

The maps list is populated when you open the Create Server dialog. Unfortunately, as usual, they programmed it in a manner that makes it restrict the path IDs for games:

```
  CModInfo* pModInfo = ModInfo();
  const char* pszGameDescription = pModInfo ->GetGameDescription();
  if ( !strcasecmp(pszGameDescription, "Half-Life") )
  {
    this->LoadMaps(nullptr);
  }
  else
  {
    this->LoadMaps("GAME");
    if ( *pModInfo->GetFallbackDir() )
    {
      if ( pModInfo->UseFallbackDirMaps() )
        this->LoadMaps("GAME_FALLBACK");
    }
  }
  this->LoadMaps("GAMEDOWNLOAD");
```

```
void CCreateMultiplayerGameServerPage::LoadMaps(const char *pszPathID)
```

If the game isn't called "Half-Life" it will only load maps from GAME and GAMEDOWNLOAD by default, and loading from GAME_FALLBACK if the fallback directory is set.

LoadMaps has additional game name checks to filter out campaign maps.

```
bool CModInfo::UseFallbackDirMaps()
{
  return strcasecmp(m_pModData->GetString( "fallback_maps", "1" ), "1") == 0;
}
```

This occurs after the client has initialized since it creates the Create Server dialog on demand (unless it forcefully creates them somehow, but i don't see that anywhere).

If it isn't listing the maps then i can only assume that either the filesystem is re-initializing itself after you've added your paths or it's skipping your paths due to using a different path ID. Using "GAME" should always work.

**Comment by tschumann** [comment]
Ah right.
I'm using GAME and adding the paths in client.dll's Initialize so I guess something else must be going on. Oh well.
With regards to checking if a user owns a game, I guess I can somehow do that on the server-side using pfnGetPlayerAuthId and ISteamUser::UserHasLicenseForApp, but pfnGetPlayerAuthId I think only returns a usable value when sv_lan is 0 - is there a better way?

**Comment by SamVanheer** [comment]
ISteamUser isn't available on the server side, it only represents a local user's data. AFAIK there is no way to check whether they own a game from that side. If you have access to the client side you can send that information to the server, though i'd be careful not to send any more data than is absolutely needed.

**Comment by tschumann** [comment]
Hm, and the server isn't available until a map loads (or at least starts loading) and I don't think there's a nice/easy way to send a message to the server after the server has started loading but before the server starts precaching. I guess the filesystem stuff was added much later so stuff like this isn't easily doable.

**Comment by tschumann** [comment]
The only other problem I encountered was maps * not returning the mounted maps (but it doesn't return what's in fallback_dir either) - anyway, I can live with that.
With regards to checking if a user owns an app, ISteamApps->BIsSubscribedApp is what you want to use. I've got a working example at https://github.com/tschumann/basis/blob/master/cl_dll/mod/steam_integration.cpp (Steam_UserOwnsApp is the function to use).

Thanks again for the help @SamVanheer.
